### PR TITLE
Column list validation before saving agent preferences (#84).

### PR DESCRIPTION
**Pull request by pboguslawski**
## Proposed change

It was noticed in OTRS 3.3, that some agents managed to save their column
filter preferences with duplikcated column names i.e.

```
    mysql -e 'SELECT * FROM user_preferences where length(preferences_value)>100' otrs
    [...]
    5       UserFilterColumnsEnabled-AgentTicketQueue       ["TicketNumber","Title","State","Lock","Queue","Owner","Age","Changed","CustomerID","Lock","Queue","CustomerID","TicketNumber", [...], "TicketNumber","Age","Title","State","Lock","Queue","CustomerID"]
```

The whole preference string was over 9kB long (all filled with duplicated
column list as above) and caused apache RAM consumption problems (removing
this record from DB was necessary to avoid out of memory on OTRS
application server). The problem was not reproducible and the cause of
the problem has not been identified (might be some kind of race condition
or web browser bug).

Same problem still exists in Znuny rel-6_0 - to simulate misbehaving
client apply patch

```diff -Nur a/var/httpd/htdocs/js/Core.Agent.Overview.js b/var/httpd/htdocs/js/Core.Agent.Overview.js
 --- a/var/httpd/htdocs/js/Core.Agent.Overview.js<->2021-05-15 16:03:56.586105278 +0200
 +++ b/var/httpd/htdocs/js/Core.Agent.Overview.js>2021-05-25 14:18:25.375359470 +0200
 @@ -83,6 +83,7 @@
                                      // if e.g. the save button is clicked multiple times
                                      if (!$('#' + Core.App.EscapeSelector(FieldID)).length) {
                                          $('<input name="UserFilterColumnsEnabled" type="hidden" />').attr('id', FieldID).val($(this).attr('data-fieldname')).appendTo($ListContainer.closest('div'));
 +                                        $('<input name="UserFilterColumnsEnabled" type="hidden" />').attr('id', FieldID).val($(this).attr('data-fieldname')).appendTo($ListContainer.closest('div'));
                                      }
                                  });
                              }
```

and change column order in Action=AgentTicketQueue in Small (S) mode.

Regardless of cause Znuny should not allow to save duplicated columns
(i.e. to prevent attacks).

This mod adds validation of column list just before saving preferences
duplicates are removed which should protect Znuny against the problem
described above.

## Type of change
- '1 - 🐞 bug 🐞'            - Bugfix (non-breaking change which fixes an issue)

## Additional information

- Related: https://github.com/OTRS/otrs/pull/1507
- Author-Change-Id: IB#1058144

## Checklist

- [x] The code change is tested and works locally.(❗)
- [x] There is no commented out code in this PR.(❕)
- [ ] You improved or added new unit tests.(❕)
- [x] Local ZnunyCodePolicy run passes successfully.(❕)
- [x] Local unit tests pass.(❕)
- [x] GitHub workflow ZnunyCodePolicy passes.(❗)
- [x] GitHub workflow unit tests pass.(❗)

### CODE REVIEW

File: Kernel/Output/HTML/Preferences/ColumnFilters.pm [hunk 1]
```diff
@@ -61,10 +62,23 @@ sub Run {
...
+            COLUMN:
+            for my $Column ( @{ $Param{GetParam}->{$Key} } ) {
+
+                # Skip duplicates.
+                next COLUMN if $Seen{$Column};
+                $Seen{$Column} = 1;
+
+                push @ColumnsUnique, $Column;
+            }
+
```
Comment (dennykorsukewitz) [inline Kernel/Output/HTML/Preferences/ColumnFilters.pm]:
```suggestion
			# remove duplicates.
            if ( $Param{GetParam}->{$Key} && IsArrayRefWithData( $Param{GetParam}->{$Key} ) ) {
                my %UniqueValues = map { $_ => 1 } @{ $Param{GetParam}->{$Key} };
                @{ $Param{GetParam}->{$Key} } = keys %UniqueValues;
            }
```

File: Kernel/Output/HTML/Preferences/ColumnFilters.pm [hunk 2]
```diff
@@ -61,10 +62,23 @@ sub Run {
...
             $Kernel::OM->Get('Kernel::System::User')->SetPreferences(
                 UserID => $Param{UserData}->{UserID},
                 Key    => $Key . '-' . $FilterAction,
-                Value  => $Kernel::OM->Get('Kernel::System::JSON')->Encode( Data => $Param{GetParam}->{$Key} ),
+                Value  => $Kernel::OM->Get('Kernel::System::JSON')->Encode( Data => \@ColumnsUnique ),
```
Comment (dennykorsukewitz) [inline Kernel/Output/HTML/Preferences/ColumnFilters.pm]:
```suggestion
                Value  => $Kernel::OM->Get('Kernel::System::JSON')->Encode( Data => $Param{GetParam}->{$Key} ),
```

File: Kernel/Output/HTML/Preferences/ColumnFilters.pm (hunk 1 above)
Comment (dennykorsukewitz) [reply Kernel/Output/HTML/Preferences/ColumnFilters.pm]: Hi @pboguslawski

these are my suggestions.
Please do not forget the following:


Line: 15
```
use Kernel::System::VariableCheck qw(:all);
```

**Review by dennykorsukewitz — CHANGES_REQUESTED** [review]
Hi @pboguslawski

these are my suggestions.

Thanks 🚀

File: Kernel/Output/HTML/Preferences/ColumnFilters.pm (hunk 1 above)
Comment (pboguslawski) [reply Kernel/Output/HTML/Preferences/ColumnFilters.pm]: Hi @dennykorsukewitz,

Your algo does not preserve item order in deduplicated array and order must be preserved in this function. Will add input type verification in next commit (today).

File: Kernel/Output/HTML/Preferences/ColumnFilters.pm (hunk 2 above)
Comment (pboguslawski) [reply Kernel/Output/HTML/Preferences/ColumnFilters.pm]: Won't be required as above.

File: Kernel/Output/HTML/Preferences/ColumnFilters.pm (hunk 1 above)
Comment (pboguslawski) [reply Kernel/Output/HTML/Preferences/ColumnFilters.pm]: Input type verification added in db23b73f13426a4dbc4948afcb7ae163ed900bba.

Comment (dennykorsukewitz) [reply Kernel/Output/HTML/Preferences/ColumnFilters.pm]: You are right. 
The order is indeed relevant at this one. 👍🏼

File: Kernel/Output/HTML/Preferences/ColumnFilters.pm [hunk 3]
```diff
@@ -57,14 +59,28 @@ sub Run {
 
     return 1 if !defined $FilterAction;
 
-    for my $Key ( sort keys %{ $Param{GetParam} } ) {
+    if ( !$Kernel::OM->Get('Kernel::Config')->Get('DemoSystem') ) {
```
Comment (dennykorsukewitz) [inline Kernel/Output/HTML/Preferences/ColumnFilters.pm:62]: 👍🏼

File: Kernel/Output/HTML/Preferences/ColumnFilters.pm [hunk 4]
```diff
@@ -57,14 +59,28 @@ sub Run {
...
+        KEY:
+        for my $Key ( sort keys %{ $Param{GetParam} } ) {
 
-        # pref update db
-        if ( !$Kernel::OM->Get('Kernel::Config')->Get('DemoSystem') ) {
+            # We require non-empty list of column names - ignore antyhing else.
+            next KEY if !( $Param{GetParam}->{$Key} && IsArrayRefWithData( $Param{GetParam}->{$Key} ) );
```
Comment (dennykorsukewitz) [inline Kernel/Output/HTML/Preferences/ColumnFilters.pm]:
```suggestion
            next KEY if !$Param{GetParam}->{$Key};
            next KEY if !IsArrayRefWithData($Param{GetParam}->{$Key});
```

File: Kernel/Output/HTML/Preferences/ColumnFilters.pm [hunk 5]
```diff
@@ -57,14 +59,28 @@ sub Run {
...
+            my %SeenColumnNames;
+            my @DeduplicatedColumnNameArray;
+            COLUMN_NAME:
+            for my $ColumnName ( @{ $Param{GetParam}->{$Key} } ) {
+                next COLUMN_NAME if $SeenColumnNames{$ColumnName};
+                $SeenColumnNames{$ColumnName} = 1;
+                push @DeduplicatedColumnNameArray, $ColumnName;
+            }
+
+            # Update list in DB.
```
Comment (dennykorsukewitz) [inline Kernel/Output/HTML/Preferences/ColumnFilters.pm]:
```suggestion
            # Remove column name duplicates from list preserving column order.
            my %SeenColumnNames;
            my @UniqueColumnNames;
            COLUMNNAME:
            for my $ColumnName ( @{ $Param{GetParam}->{$Key} } ) {
                next COLUMNNAME if $SeenColumnNames{$ColumnName};
                $SeenColumnNames{$ColumnName} = 1;
                push @UniqueColumnNames, $ColumnName;
            }

```

File: Kernel/Output/HTML/Preferences/ColumnFilters.pm [hunk 6]
```diff
@@ -57,14 +59,28 @@ sub Run {
...
             $Kernel::OM->Get('Kernel::System::User')->SetPreferences(
                 UserID => $Param{UserData}->{UserID},
                 Key    => $Key . '-' . $FilterAction,
-                Value  => $Kernel::OM->Get('Kernel::System::JSON')->Encode( Data => $Param{GetParam}->{$Key} ),
+                Value  => $Kernel::OM->Get('Kernel::System::JSON')->Encode( Data => \@DeduplicatedColumnNameArray ),
```
Comment (dennykorsukewitz) [inline Kernel/Output/HTML/Preferences/ColumnFilters.pm]:
```suggestion
                Value  => $Kernel::OM->Get('Kernel::System::JSON')->Encode( Data => \@UniqueColumnNames ),
```

**Review by dennykorsukewitz — CHANGES_REQUESTED** [review]

**Comment by dennykorsukewitz** [comment]
Hi @pboguslawski

here are my new suggestions after your hint. 

Thanks 🚀

File: Kernel/Output/HTML/Preferences/ColumnFilters.pm (hunk 4 above)
Comment (pboguslawski) [reply Kernel/Output/HTML/Preferences/ColumnFilters.pm]: Done in 4252343ca6f6cdf84d540a9c08635c414511d75b.

File: Kernel/Output/HTML/Preferences/ColumnFilters.pm (hunk 5 above)
Comment (pboguslawski) [reply Kernel/Output/HTML/Preferences/ColumnFilters.pm]: Done in 4252343ca6f6cdf84d540a9c08635c414511d75b.

File: Kernel/Output/HTML/Preferences/ColumnFilters.pm (hunk 6 above)
Comment (pboguslawski) [reply Kernel/Output/HTML/Preferences/ColumnFilters.pm]: Done in 4252343ca6f6cdf84d540a9c08635c414511d75b.

**Comment by pboguslawski** [comment]
> here are my new suggestions after your hint.

Done in 4252343ca6f6cdf84d540a9c08635c414511d75b.

**Review by dennykorsukewitz — APPROVED** [review]

**Review by dennykorsukewitz — APPROVED** [review]

**Pull request merged**